### ISSUE TITLE
Cannot use nix-env from regular user

### ISSUE THREAD

**Issue by joepie91** [NixOS/nixpkgs#12399]
Like #5713, I was unable to install packages from my limited user. However, I am using Bash, not zsh. My NixOS version is `15.09.911.53286a8 (Dingo)`.

```
[sven@nixos:~]$ nix-env -i openttd
error: Nix database directory ���/nix/var/nix/db��� is not writable: Permission denied
```

(Ignore the mojibake, I'm having some trouble with my local shell - this is in an SSH session.)

I followed the suggested workaround by specifying `export NIX_REMOTE=daemon`, which worked. I also attempted to set `programs.bash.enable = true;`, which resulted in this message after a rebuild:

```
warning: The option `programs.bash.enable' defined in your configuration no longer has any effect; please remove it.
```

However, no `.bashrc`, `.bash_profile` or `.profile` was generated for my user *at all*, with or without the setting. The following is my `extraUsers` configuration in `/etc/nixos/configuration.nix`:

```nix
  users.extraUsers.sven = {
    isNormalUser = true;
    uid = 1000;
    openssh.authorizedKeys.keys = ["<snip>"];
  };
```

Is there something I'm missing here, or is this actually a bug?

EDIT: To clarify, I've SSH'ed in as root, and did `su sven` to drop down to the user.

**Comment by vcunat** [comment]
> To clarify, I've SSH'ed in as root, and did su sven to drop down to the user.

That seems to be the problem. When I `su myUser` from root, the env variables aren't correctly set up; in particular, `$NIX_REMOTE` is missing. With normal login all works fine for me.

**Comment by vcunat** [comment]
I don't really know these commands properly, but perhaps one's supposed to use `su -l myUser` instead. That fixes the problem for me.

**Comment by joepie91** [comment]
Shouldn't it still use a `.profile`, even with a login shell? There don't seem to be any generated user configuration files *at all* for my user. Either way it should probably generate this for .bashrc as well, if it doesn't do so already.

**Comment by vcunat** [comment]
There are some `/etc/*` files sourced IIRC.

**Comment by jgillich** [comment]
FWIW, I just tried this with fish as my user shell, and it works fine.

**Comment by joepie91** [comment]
Right. I understand that using a login shell works around it, but this should really work without a login shell as well - is there a similar file in `/etc` for the `.bashrc` (ie. non-login shells), perhaps?

**Comment by peti** [comment]
> I understand that using a login shell works around it, but this should really work without a login shell as well

That error has nothing to do with whether you're using a login shell or not. The issue is that your environment isn't set up properly and therefore your `nix-env` tries to access files directly instead of going through `nix-daemon`, like it should. You can probably remedy this issue by running "`source /etc/profile`".

Anyhow, when you log in as `root`, then you'll get an environment in which Nix can be run as `root`. It seems unrealistic to expect that you can switch to a different user without fixing your environment and have Nix still work properly like before.

**Comment by joepie91** [comment]
Isn't the purpose of the configuration in `/etc/profile` to make Nix function through the daemon regardless of what user account is being used? Given that, it makes no sense to make it depend on *how* the user account is accessed.

If NixOS is to provide a consistent user experience, then the following two scenarios should both work correctly out of the box:

1. Log in directly to a user (whether over SSH or on a local terminal)
2. Log in to another user, and `su` to the target user

Right now, only scenario 1 is supported. What technical reason is there not to make scenario 2 behave the same, especially given that this is a fairly common pattern on multi-user systems?

And yes, it absolutely has to do with it using a login shell - `/etc/profile` is only executed for login shells. It even says so in the file itself. For non-login shells,`/etc/bashrc` is run (which also exists, but does not set `NIX_REMOTE`).

In short: "the environment isn't set up properly" should not be my problem as an end user. It's not transparent to me how or why it works in one case but not the other, so why would I (or any other end user) be expected to fix it? This is something that needs to work out of the box to make NixOS' abstraction hold up.

**Comment by peti** [comment]
I don't know ... I don't believe that logging in as root and then su'ing to another user is a "common pattern".  I certainly don't work that way. I would say that the common pattern is to do the exact opposite, i.e. most people seem to log in as a normal user and then su to become root when they need to.

**Comment by joepie91** [comment]
It is a very common when configuring things for other users on a system that you have root access to. Regardless, it should be a *supported* pattern - not every shell is a login shell.

**Comment by peti** [comment]
So you seem to be saying that there should be no differences between login shells and non-login shells, right? Because you *don't* want to request a login shell, but you want everything to work anyway. Clearly that expectation cannot be met if the environment of login shells differs from non-login shells in any way.

**Comment by joepie91** [comment]
I expect things to work in a non-login shell like they work in a login shell, *if* there is no technical reason for them to differ. The `NIX_REMOTE` variable is one of those cases - it clearly works fine in a non-login shell, so why shouldn't it be set by default?

**Comment by peti** [comment]
I was asking a very specific question about a very specific thing, i.e. the shell environment. Would you say that the environment defined by a non-login shell must always be identical to that defined by a login shell?

**Comment by joepie91** [comment]
If there are no technical reasons for there to be a difference, then sure. I have no idea whether there are.

**Comment by peti** [comment]
I am sorry, but I don't feel like I'm making progress understanding your POV. The question I asked can be answered with *yes* or *no*, but I feel like I'm not getting either answer.

Anyway, my expectation is that "su username" changes the uid to that user but leaves my current shell environment untouched. If I want the shell environment to change, then I run "su - username". This approach has worked just fine for me for 20+ years of working on Unix, and NixOS is no exception there. If you'd like different semantics implemented, then maybe it's best to propose a concrete change by means of a PR.
